Compatibility with numpy 2

arunkannawadi opened the pull request:
This PR makes `ngmix` compatible with `numpy>=2` and `numpy<2`.

The recommended method is to replace `np.array(..., copy=False)` with `np.asarray` [link](https://numpy.org/devdocs/numpy_2_0_migration_guide.html#adapting-to-changes-in-the-copy-keyword). However, given the necessity of `ndmin` which is not available in `asarray`, I took the [scipy approach](https://github.com/scipy/scipy/pull/20172) here.

arunkannawadi committed:
BUG Cast to np.uint32

When the input variable ``val`` may be of type np.int32. This forces an
implicit casting of ``fval``, and when it is equal to or exceed 2**31,
an OverflowError is raised. This explicit casting prevents this error.

arunkannawadi committed:
TST Add py3.12 to the matrix

arunkannawadi committed:
ENH Make ngmix compatabile across numpy 2

This adds a shim similar to scipy that sets copy=False in numpy < 2 and
copy=None otherwise.

arunkannawadi committed:
PROD Update the build status badge

arunkannawadi commented on .github/workflows/test.yml:
```diff
@@ -11,7 +11,7 @@ jobs:
     strategy:
       matrix:
-        pyver: ["3.9", "3.10", "3.11"]
+        pyver: ["3.9", "3.10", "3.11", "3.12"]
```
I added 3.12 to explicitly check this with the python version in the latest `rubin-env`. I left 3.9 as is, since it acts as an implicit check with `numpy<2`.

beckermr replied:
SGTM!

arunkannawadi commented on ngmix/defaults.py:
```diff
@@ -1,5 +1,10 @@
+if np.lib.NumpyVersion(np.__version__) >= "2.0.0":
+    copy_if_needed = None
+else:
+    copy_if_needed = False
```
`defaults` may not be the best place for this, but happy to move it wherever it might be more obvious for you.

arunkannawadi commented on ngmix/flags.py:
```diff
@@ -77,7 +79,7 @@ def get_flags_str(val, name_map=None):
     for pow in range(32):
         fval = 2**pow
-        if ((val & fval) != 0):
+        if ((np.uint32(val) & fval) != 0):
```
With `numpy>=2`, this now results in an OverflowError. We could either adopt limiting `pow` to `range(30)` as we did in [metadetect here](https://github.com/esheldon/metadetect/pull/224). I figured it's probably safer to cast it instead since that leaves the most-significant bits still usable.

beckermr replied:
We should 

- limit the loop to 31
- raise a warning if the input value is bigger than 2**32-1 or < 0
- safely cast using `np.array(value).astype(dtype)`

A direct call to `np.uint32` is raising a warning that it will raise an error in the future for my local copy of numpy at 1.26.4.

arunkannawadi replied:
Oh, I didn't know there were going to be future releases of `numpy 1.26` that would convert this to an error.

beckermr replied:
They might mean in numpy 2. IDK.

arunkannawadi committed:
DOC Specify the range of flag bits

arunkannawadi committed:
BUG Safely typecast val

arunkannawadi committed:
ENH Warn if val is outside the range

beckermr commented on ngmix/flags.py:
```diff
@@ -74,10 +76,14 @@ def get_flags_str(val, name_map=None):
+    for pow in range(31):
         fval = 2**pow
-        if ((val & fval) != 0):
+        if ((np.array(val).astype(np.uint32) & fval) != 0):
```
Let's move this cast up to the top and do it once.

arunkannawadi replied:
I want to make sure we don't do too many redundant things. If we limit the loop to `range(31)`, then casting isn't even required. The error was arising from implicit casting of `fval` and not `val`. We should perhaps raise a `ValueError` if `val` is outside the specific range to avoid any underflow errors be done with it. For instance, if `val = np.int32(-5)`, we'd get a completely wrong result with just a seemingly benign warning.

arunkannawadi replied:
My suggestion is to modify your list above to the following:

We should

- limit the loop to 31
- raise an exception if the input value is bigger than 2**32-1 or < 0

arunkannawadi replied:
OK, in that case, having an upper limit doesn't make sense then. We raise an exception if `val` is negative and cast it to `uint64` safely always.

arunkannawadi replied:
Well, what do you know. The code will now work even if `val` is `np.int8` etc.

arunkannawadi replied:
My plan of always casting to `uint64` doesn't work for `numpy<2`. However, given the range of values that `fval` takes, which is hard coded in this routine,`uint32` should be sufficient. It does not pose any restriction on the exact integer type of `val` .

arunkannawadi replied:
I'd hate to break anything because I didn't consider a corner case, so I added a small unit test as well.

arunkannawadi committed:
BUG Cast to unit64 for maximal flexbility

arunkannawadi committed:
Fall back to casting to uint32

With numpy<2, bitwise_and does not operate on uint64 dtype objects.
Given that the range of fval is within the range of uint32, casting to
uint32 suffices. If the input value has more bits, they will be not used
anyway regardless of the casting as long as fval is limited to 2**31.

arunkannawadi committed:
TST Check that get_flags_str works for different int types

arunkannawadi committed:
STY Please flake8

beckermr commented on ngmix/tests/test_flags.py:
```diff
@@ -0,0 +1,14 @@
+)
+def test_get_flags_str(dtype):
+    """Test that get_flags_str works for different integer types."""
+    val = np.array(LOW_DET, dtype=dtype)
```
For the bigger types we should or with 2**45 to ensure things are OK.

arunkannawadi replied:
In `numpy<2`, bitwise or to `uint64` is prohibited for some reason. I'll condition on that if that's okay.

beckermr replied:
Ack I forgot and yes pls do!

arunkannawadi committed:
TST OR with a large bit

beckermr committed:
Fix test_get_flags_str for large integers

arunkannawadi committed:
TST Exempt setting bit 45 for uint64

arunkannawadi committed:
TST Use astype to avoid numpy complaining

arunkannawadi commented on ngmix/tests/test_flags.py:
```diff
@@ -11,8 +11,6 @@
-    val = np.array(LOW_DET, dtype=dtype)
-    if val.itemsize > 4 and dtype not in (np.uint64, ):
-        val |= 2**45
+    val = np.array(LOW_DET | 2**45).astype(dtype)
```
Alright, this looks much cleaned than the test with all the conditions.

beckermr replied:
Yeah it just leaves the question of how numpy handles truncating things. That's ok I guess but maybe a comment would be good.

arunkannawadi committed:
[ci skip] DOC Comment on truncation with astype

beckermr commented:
@esheldon comments here or shall we merge?

arunkannawadi committed:
TST Check that comment on truncation is correct

esheldon commented:
Arun are you still making additions?

arunkannawadi commented:
Nope, I don't have any more to add unless there are requests. Could you also please cut a release after merging this?

arunkannawadi commented:
I shall, but I also notice that changes from #246 have not made it into a release yet. I will let you comment on that. This would bump up the minor version, right?

arunkannawadi committed:
[ci skip] DOC Edit the comment to cast to unint32

esheldon commented:
I'll add notes for that.  I'll do the version bump when I make my changes.

arunkannawadi committed:
[ci skip] MAINT Add changes to CHANGES.md

esheldon commented:
v2.4.0 is released

arunkannawadi commented:
Thank you, Erin!
